False positive for ``super-init-not-called``

### Bug description

```python
class A:
    def __init__(self):
        pass


class B(A):
    pass


class D(B, A):
    def __init__(self):
        A.__init__(self)
```

This was found while working on #4940, after fixing this we might want to check if there are any messages that are unnecessarily disabled in our own codebase.

### Command used

```shell
pylint test.py
```


### Pylint output

```shell
pylint test.py
************* Module test
test.py:11:4: W0231: __init__ method from base class 'B' is not called (super-init-not-called)

------------------------------------------------------------------
Your code has been rated at 8.75/10 (previous run: 8.75/10, +0.00)
```


### Expected behavior

As class `B` does not have a `__init__` it can't be called. The only `__init__` included in class `B` is the `__init__` of class `A` which is being called in class `C`.

### Pylint version

```shell
pylint 2.10.3-dev0
astroid 2.7.3
Python 3.9.6 (default, Aug  2 2021, 08:23:35) 
[Clang 12.0.5 (clang-1205.0.22.11)]
```


### OS / Environment

macOS

### Additional dependencies

_No response_

--- a/ChangeLog
+++ b/ChangeLog
@@ -215,6 +215,11 @@ Release date: TBA
   Closes #4434
   Closes #5370
 
+* Fix false positive ``super-init-not-called`` for classes that inherit their ``init`` from
+  a parent.
+
+  Closes #4941
+
 * ``encoding`` can now be supplied as a positional argument to calls that open
   files without triggering ``unspecified-encoding``.
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -124,6 +124,11 @@ Other Changes
 
   Closes #5483
 
+* Fix false positive ``super-init-not-called`` for classes that inherit their ``init`` from
+  a parent.
+
+  Closes #4941
+
 * Fix ``unnecessary_dict_index_lookup`` false positive when deleting a dictionary's entry.
 
   Closes #4716

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -52,10 +52,10 @@
 """Classes checker for Python code"""
 import collections
 from itertools import chain, zip_longest
-from typing import List, Pattern
+from typing import List, Pattern, Set
 
 import astroid
-from astroid import nodes
+from astroid import bases, nodes
 
 from pylint.checkers import BaseChecker, utils
 from pylint.checkers.utils import (
@@ -1939,6 +1939,7 @@ a metaclass class method.",
         klass_node = node.parent.frame(future=True)
         to_call = _ancestors_to_call(klass_node)
         not_called_yet = dict(to_call)
+        parents_with_called_inits: Set[bases.UnboundMethod] = set()
         for stmt in node.nodes_of_class(nodes.Call):
             expr = stmt.func
             if not isinstance(expr, nodes.Attribute) or expr.attrname != "__init__":
@@ -1971,7 +1972,9 @@ a metaclass class method.",
                     if isinstance(klass, astroid.objects.Super):
                         return
                     try:
-                        del not_called_yet[klass]
+                        method = not_called_yet.pop(klass)
+                        # Record that the class' init has been called
+                        parents_with_called_inits.add(node_frame_class(method))
                     except KeyError:
                         if klass not in to_call:
                             self.add_message(
@@ -1980,6 +1983,11 @@ a metaclass class method.",
             except astroid.InferenceError:
                 continue
         for klass, method in not_called_yet.items():
+            # Check if the init of the class that defines this init has already
+            # been called.
+            if node_frame_class(method) in parents_with_called_inits:
+                return
+
             if decorated_with(node, ["typing.overload"]):
                 continue
             cls = node_frame_class(method)

--- a/pylint/checkers/imports.py
+++ b/pylint/checkers/imports.py
@@ -427,9 +427,7 @@ class ImportsChecker(DeprecatedMixin, BaseChecker):
         ),
     )
 
-    def __init__(
-        self, linter: Optional["PyLinter"] = None
-    ):  # pylint: disable=super-init-not-called # See https://github.com/PyCQA/pylint/issues/4941
+    def __init__(self, linter: Optional["PyLinter"] = None) -> None:
         BaseChecker.__init__(self, linter)
         self.import_graph: collections.defaultdict = collections.defaultdict(set)
         self._imports_stack: List[Tuple[Any, Any]] = []

--- a/pylint/checkers/stdlib.py
+++ b/pylint/checkers/stdlib.py
@@ -463,9 +463,7 @@ class StdlibChecker(DeprecatedMixin, BaseChecker):
         ),
     }
 
-    def __init__(
-        self, linter: Optional["PyLinter"] = None
-    ):  # pylint: disable=super-init-not-called # See https://github.com/PyCQA/pylint/issues/4941
+    def __init__(self, linter: Optional["PyLinter"] = None) -> None:
         BaseChecker.__init__(self, linter)
         self._deprecated_methods: Set[Any] = set()
         self._deprecated_methods.update(DEPRECATED_METHODS[0])

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -1,7 +1,7 @@
 """This should not emit a super-init-not-called warning. It previously did this, because
 ``next(node.infer())`` was used in that checker's logic and the first inferred node
 was an Uninferable object, leading to this false positive."""
-# pylint: disable=too-few-public-methods
+# pylint: disable=too-few-public-methods, missing-class-docstring
 
 import ctypes
 
@@ -11,3 +11,31 @@ class Foo(ctypes.BigEndianStructure):
 
     def __init__(self):
         ctypes.BigEndianStructure.__init__(self)
+
+
+# Tests for not calling the init of a parent that does not define one
+# but inherits it.
+class GrandParentWithInit:
+    def __init__(self):
+        print(self)
+
+
+class ParentWithoutInit(GrandParentWithInit):
+    pass
+
+
+class ChildOne(ParentWithoutInit, GrandParentWithInit):
+    """Since ParentWithoutInit calls GrandParentWithInit it doesn't need to be called."""
+
+    def __init__(self):
+        GrandParentWithInit.__init__(self)
+
+
+class ChildTwo(ParentWithoutInit):
+    def __init__(self):
+        ParentWithoutInit.__init__(self)
+
+
+class ChildThree(ParentWithoutInit):
+    def __init__(self):  # [super-init-not-called]
+        ...

--- a/tests/functional/s/super/super_init_not_called.txt
+++ b/tests/functional/s/super/super_init_not_called.txt
@@ -1,0 +1,1 @@
+super-init-not-called:40:4:41:11:ChildThree.__init__:__init__ method from base class 'ParentWithoutInit' is not called:UNDEFINED

